src/riak_kv_qry.erl

```diff
@@ -180,18 +190,29 @@ xlate_insert_to_putdata(Values, Positions, Empty) ->
             {error, {too_many_insert_values, lists:reverse(Errors)}}
     end.
 
--spec make_insert_row([] | [riak_ql_ddl:data_value()], [] | [pos_integer()], tuple()) ->
+-spec make_insert_row([riak_ql_ddl:data_value()], [pos_integer()], tuple(),
+                      [riak_ql_ddl:simple_field_type()]) ->
                       {ok, tuple()} | {error, string()}.
-make_insert_row([], _Positions, Row) when is_tuple(Row) ->
+make_insert_row(Vals, _Positions, Row, _FieldTypes)
+  when length(Vals) > size(Row) ->
+    %% diagnose too_many_values before eventual timestamp conversion
+    %% errors, so that it is reported with higher precedence than that
+    {error, too_many_values};
+make_insert_row([], _Positions, Row, _FieldTypes) ->
     %% Out of entries in the value - row is populated with default values
     %% so if we run out of data for implicit/explicit fieldnames can just return
     {ok, Row};
-make_insert_row(_, [], Row) when is_tuple(Row) ->
-    %% Too many values for the field
-    {error, too_many_values};
-%% Make sure the types match
-make_insert_row([{_Type, Val} | Values], [Pos | Positions], Row) when is_tuple(Row) ->
-    make_insert_row(Values, Positions, setelement(Pos, Row, Val)).
+make_insert_row([TypedVal | Values], [Pos | Positions], Row, FieldTypes) ->
```

Comment:
It's interesting but tricky. For one thing, as used here, the row may be incomplete and the argument values may come in arbitrary order (which happens when INSERT comes with field names). Or perhaps your suggestion is to have the `Mod:maybe_convert_timestamp` take the full row?

I'm still reluctant to take it on, at least now, because doing so would raise the demons of DDL version change and new capability handling. Also, calling an external, non-OTP function from a helper module is something I can't quite see the full consequences of.